Description: Build the MATLAB mex files without license management
 Upstream's MATLAB build links every PTB module against libLexActivator from
 PsychPlugins (deleted in postPatch). The C side already has a no-license
 path: PsychLM.c and its external LexActivator.h include are inside
 #ifndef PSYCH_NOLM, and MiniBox.c has stubs under #ifdef PSYCH_NOLM. So this
 patch only changes the build script:
 - Add -DPSYCH_NOLM. Remove -lLexActivator and the -L and $ORIGIN rpath to
   PsychPlugins, which exist only for that library.
 - Get include paths from pkg-config instead of /usr/include and Debian
   multiarch paths, as use-pkg-config.patch does for the Octave script.
 - Use the system GLEW (-lGLEW, no -DGLEW_STATIC, no glew.c), as
   use-system-glew.patch does. Drop -DPTB_USE_NVSTUSB, as no-nvstusb.patch
   does.
 - Build PsychCV without the /usr/include/apriltag probe and without
   -lapriltag-utils, and Eyelink with the include path from pkg-config.
 - Add -Wno-address and -Wno-int-conversion to moglcore, as the Octave
   script has. GCC 14 and later make int-conversion an error.
 - Skip modes 7, 12 and 14 (libfreenect, Oculus SDK, OpenHMD). Their blocks
   are unchanged and still link libLexActivator.
 The mex calls that change use function syntax, so that the pkg-config flags
 can be spliced in.
 Apply after the Octave patches in package.nix; the C changes of
 use-system-glew.patch (#include <GL/glew.h>) are needed here too.
Forwarded: not-needed
---
--- a/PsychSourceGL/Source/linuxmakeit64.m
+++ b/PsychSourceGL/Source/linuxmakeit64.m
@@ -27,30 +27,43 @@
     return;
 end
 
+% Nix: the builder supplies no libfreenect (7), Oculus SDK (12) or
+% OpenHMD (14), and the blocks below still link libLexActivator.
+if any(mode == [7 12 14])
+    fprintf('Skipping plugin type %i: not supported by the Nix build.\n\n', mode);
+    return;
+end
+
 fprintf('Building plugin type %i ...\n\n', mode);
 
+% Nix: all PTB modules are built with -DPSYCH_NOLM and without
+% -lLexActivator, -L/-rpath PsychPlugins (only libLexActivator lives there),
+% hardcoded /usr/include paths, or the embedded GLEW. The Nix builder sets
+% the rpath after the build.
+
 if mode==0
     % Build Screen.mexa64: We build against GStreamer-1
     mex -c Common/Screen/tinyexr.cc
-    mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread' -outdir ../Projects/Linux/build/ -output Screen -largeArrayDims -DMEX_DOUBLE_HANDLE -DPTBMODULE_Screen -DPTB_USE_GSTREAMER -DPTBVIDEOCAPTURE_LIBDC -DPTB_USE_NVSTUSB -DGLEW_STATIC -D_GNU_SOURCE -I/usr/include -I/usr/include/gstreamer-1.0 -I/usr/lib/x86_64-linux-gnu/gstreamer-1.0/include -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/lib/i386-linux-gnu/glib-2.0/include -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/libxml2 -I../Cohorts/libnvstusb-code-32/include -ICommon/Base -ICommon/Screen -ILinux/Base -ILinux/Screen "Linux/Base/*.c" "Linux/Screen/*.c" "Common/Screen/*.c" "Common/Base/*.c" tinyexr.o -lc -lLexActivator -lrt -ldl -lGL -lGLU -lX11 -lXext -lX11-xcb -lxcb -lxcb-dri3 -lxcb-present -lgstreamer-1.0 -lgstbase-1.0 -lgstapp-1.0 -lgstvideo-1.0 -lgstpbutils-1.0 -lgobject-2.0 -lgmodule-2.0 -lxml2 -lgthread-2.0 -lglib-2.0 -lXxf86vm -ldc1394 -lusb-1.0 -lpciaccess -lXi -lXrandr -lXfixes -lXcomposite
+    iflags = pkgconfig_iflags('gstreamer-1.0 gstreamer-app-1.0 gstreamer-video-1.0 gstreamer-pbutils-1.0 glib-2.0 libxml-2.0');
+    mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread', '-outdir', '../Projects/Linux/build/', '-output', 'Screen', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-DPTBMODULE_Screen', '-DPTB_USE_GSTREAMER', '-DPTBVIDEOCAPTURE_LIBDC', '-D_GNU_SOURCE', '-ICommon/Base', '-ICommon/Screen', '-ILinux/Base', '-ILinux/Screen', iflags{:}, 'Linux/Base/*.c', 'Linux/Screen/*.c', 'Common/Screen/*.c', 'Common/Base/*.c', 'tinyexr.o', '-lc', '-lrt', '-ldl', '-lGL', '-lGLU', '-lGLEW', '-lX11', '-lXext', '-lX11-xcb', '-lxcb', '-lxcb-dri3', '-lxcb-present', '-lgstreamer-1.0', '-lgstbase-1.0', '-lgstapp-1.0', '-lgstvideo-1.0', '-lgstpbutils-1.0', '-lgobject-2.0', '-lgmodule-2.0', '-lxml2', '-lgthread-2.0', '-lglib-2.0', '-lXxf86vm', '-ldc1394', '-lusb-1.0', '-lpciaccess', '-lXi', '-lXrandr', '-lXfixes', '-lXcomposite');
     unix(['mv ../Projects/Linux/build/Screen.' mexext ' ' PsychtoolboxRoot 'PsychBasic/']);
 end
 
 if mode==1
     % Build GetSecs.mexa64:
-    mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread' -outdir ../Projects/Linux/build/ -output GetSecs -largeArrayDims -DMEX_DOUBLE_HANDLE -DPTBMODULE_GetSecs -ICommon/Base -ILinux/Base -ICommon/GetSecs -ICommon/Screen  "Linux/Base/*.c" "Common/Base/*.c" "Common/GetSecs/*.c" -lc -lLexActivator -lrt -ldl
+    mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread', '-outdir', '../Projects/Linux/build/', '-output', 'GetSecs', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-DPTBMODULE_GetSecs', '-ICommon/Base', '-ILinux/Base', '-ICommon/GetSecs', '-ICommon/Screen', 'Linux/Base/*.c', 'Common/Base/*.c', 'Common/GetSecs/*.c', '-lc', '-lrt', '-ldl');
     unix(['mv ../Projects/Linux/build/GetSecs.' mexext ' ' PsychtoolboxRoot 'PsychBasic/']);
 end
 
 if mode==2
     % Build WaitSecs.mexa64:
-    mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread' -outdir ../Projects/Linux/build/ -output WaitSecs -largeArrayDims -DMEX_DOUBLE_HANDLE -DPTBMODULE_WaitSecs -ICommon/Base -ILinux/Base -ICommon/WaitSecs -ICommon/Screen  "Linux/Base/*.c" "Common/Base/*.c" "Common/WaitSecs/*.c" -lc -lLexActivator -lrt -ldl
+    mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread', '-outdir', '../Projects/Linux/build/', '-output', 'WaitSecs', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-DPTBMODULE_WaitSecs', '-ICommon/Base', '-ILinux/Base', '-ICommon/WaitSecs', '-ICommon/Screen', 'Linux/Base/*.c', 'Common/Base/*.c', 'Common/WaitSecs/*.c', '-lc', '-lrt', '-ldl');
     unix(['mv ../Projects/Linux/build/WaitSecs.' mexext ' ' PsychtoolboxRoot 'PsychBasic/']);
 end
 
 if mode==3
     % Build PsychPortAudio.mexa64:
-    mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread' -outdir ../Projects/Linux/build/ -output PsychPortAudio -largeArrayDims -DMEX_DOUBLE_HANDLE -DPTBMODULE_PsychPortAudio -ICommon/Base -ILinux/Base -ICommon/PsychPortAudio -ICommon/Screen  "Linux/Base/*.c" "Common/Base/*.c" "Common/PsychPortAudio/*.c" -lportaudio -lasound -lc -lLexActivator -lrt -ldl
+    mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread', '-outdir', '../Projects/Linux/build/', '-output', 'PsychPortAudio', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-DPTBMODULE_PsychPortAudio', '-ICommon/Base', '-ILinux/Base', '-ICommon/PsychPortAudio', '-ICommon/Screen', 'Linux/Base/*.c', 'Common/Base/*.c', 'Common/PsychPortAudio/*.c', '-lportaudio', '-lasound', '-lc', '-lrt', '-ldl');
     unix(['mv ../Projects/Linux/build/PsychPortAudio.' mexext ' ' PsychtoolboxRoot 'PsychBasic/']);
 end
 
@@ -58,7 +71,8 @@
     % Build Eyelink.mexa64: Needs Eyelink SDK (eyelink_core component) from SR-Research.
     % See https://www.sr-research.com/support/docs.php?topic=linuxsoftware for install instructions.
     try
-    mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread -Wno-deprecated-declarations' -outdir ../Projects/Linux/build/ -output Eyelink -largeArrayDims -DMEX_DOUBLE_HANDLE -DPTBMODULE_Eyelink -I/usr/include/EyeLink -ICommon/Base -ILinux/Base -ICommon/Eyelink -ICommon/Screen  "Linux/Base/*.c" "Common/Base/*.c" "Common/Eyelink/*.c" -leyelink_core -lc -lrt -ldl
+    iflags = pkgconfig_iflags('eyelink_core');
+    mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread -Wno-deprecated-declarations', '-outdir', '../Projects/Linux/build/', '-output', 'Eyelink', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-DPTBMODULE_Eyelink', iflags{:}, '-ICommon/Base', '-ILinux/Base', '-ICommon/Eyelink', '-ICommon/Screen', 'Linux/Base/*.c', 'Common/Base/*.c', 'Common/Eyelink/*.c', '-leyelink_core', '-lc', '-lrt', '-ldl');
     unix(['mv ../Projects/Linux/build/Eyelink.' mexext ' ' PsychtoolboxRoot 'PsychBasic/']);
     catch
         disp(psychlasterror);
@@ -67,7 +81,7 @@
 
 if mode==5
     % Build IOPort.mexa64:
-    mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread' -outdir ../Projects/Linux/build/ -output IOPort -largeArrayDims -DMEX_DOUBLE_HANDLE -DPTBMODULE_IOPort -ICommon/Base -ILinux/Base -ICommon/IOPort -ICommon/Screen  "Linux/Base/*.c" "Common/Base/*.c" "Common/IOPort/*.c" -lc -lLexActivator -lrt -ldl
+    mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread', '-outdir', '../Projects/Linux/build/', '-output', 'IOPort', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-DPTBMODULE_IOPort', '-ICommon/Base', '-ILinux/Base', '-ICommon/IOPort', '-ICommon/Screen', 'Linux/Base/*.c', 'Common/Base/*.c', 'Common/IOPort/*.c', '-lc', '-lrt', '-ldl');
     unix(['mv ../Projects/Linux/build/IOPort.' mexext ' ' PsychtoolboxRoot 'PsychBasic/']);
 end
 
@@ -76,7 +90,7 @@
     curdir = pwd;
     cd('../../Psychtoolbox/PsychOpenGL/MOGL/source/')
     try
-    mex CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread -Wno-discarded-qualifiers -Wno-incompatible-pointer-types' -outdir ./ -output moglcore -DLINUX -DGLEW_STATIC -largeArrayDims -DMEX_DOUBLE_HANDLE -I./ -I/usr/include -lc -lGL -lGLU -lglut moglcore.c gl_auto.c gl_manual.c glew.c mogl_rebinder.c ftglesGlue.c
+    mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread -Wno-address -Wno-discarded-qualifiers -Wno-incompatible-pointer-types -Wno-int-conversion', '-outdir', './', '-output', 'moglcore', '-DLINUX', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-I./', '-lc', '-lGL', '-lGLU', '-lglut', '-lGLEW', 'moglcore.c', 'gl_auto.c', 'gl_manual.c', 'mogl_rebinder.c', 'ftglesGlue.c');
     catch %#ok<*CTCH>
     end
     unix(['mv moglcore.' mexext ' ' PsychtoolboxRoot 'PsychBasic/']);
@@ -91,7 +105,8 @@
 
 if mode==8
     % Build PsychHID.mexa64:
-    mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread' -outdir ../Projects/Linux/build/ -output PsychHID -largeArrayDims -DMEX_DOUBLE_HANDLE -DPTBMODULE_PsychHID -I/usr/include/libusb-1.0 -ICommon/Base -ILinux/Base -ICommon/PsychHID -ILinux/PsychHID -ICommon/Screen  "Linux/Base/*.c" "Common/Base/*.c" "Common/PsychHID/*.c" "Linux/PsychHID/*.c" -lc -lLexActivator -ldl -lrt -lusb-1.0 -lX11 -lXi -lutil
+    iflags = pkgconfig_iflags('libusb-1.0');
+    mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread', '-outdir', '../Projects/Linux/build/', '-output', 'PsychHID', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-DPTBMODULE_PsychHID', iflags{:}, '-ICommon/Base', '-ILinux/Base', '-ICommon/PsychHID', '-ILinux/PsychHID', '-ICommon/Screen', 'Linux/Base/*.c', 'Common/Base/*.c', 'Common/PsychHID/*.c', 'Linux/PsychHID/*.c', '-lc', '-ldl', '-lrt', '-lusb-1.0', '-lX11', '-lXi', '-lutil');
     unix(['mv ../Projects/Linux/build/PsychHID.' mexext ' ' PsychtoolboxRoot 'PsychBasic/']);
 end
 
@@ -107,9 +122,12 @@
     cd(curdir);
 end
 
-if mode == 10 && exist('/usr/include/apriltag', 'dir')
-    % Build PsychCV.mexa64:
-    mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread' -outdir ../Projects/Linux/build/ -output PsychCV -largeArrayDims -DMEX_DOUBLE_HANDLE -DPTBMODULE_PsychCV -DPSYCHCV_USE_APRILTAGS -I/usr/include/apriltag -ICommon/Base -ILinux/Base -ICommon/PsychCV "Common/Base/*.c" "Linux/Base/*.c" "Common/PsychCV/*.c" -lc -lLexActivator -lrt -ldl -lapriltag -lapriltag-utils
+if mode == 10
+    % Build PsychCV.mexa64. Upstream apriltag builds the common/ helpers into
+    % libapriltag itself (no libapriltag-utils), and PsychCV includes its
+    % headers without the apriltag/ prefix.
+    iflags = strcat(pkgconfig_iflags('apriltag'), '/apriltag');
+    mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread', '-outdir', '../Projects/Linux/build/', '-output', 'PsychCV', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-DPTBMODULE_PsychCV', '-DPSYCHCV_USE_APRILTAGS', iflags{:}, '-ICommon/Base', '-ILinux/Base', '-ICommon/PsychCV', 'Common/Base/*.c', 'Linux/Base/*.c', 'Common/PsychCV/*.c', '-lc', '-lrt', '-ldl', '-lapriltag');
     unix(['mv ../Projects/Linux/build/PsychCV.' mexext ' ' PsychtoolboxRoot 'PsychBasic/']);
 end
 
@@ -149,7 +167,7 @@
     % Build PsychVulkanCore.mexa64:
     % Needs at least Vulkan SDK version 1.2.189
     try
-        mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread' -outdir ../Projects/Linux/build/ -output PsychVulkanCore -largeArrayDims -DMEX_DOUBLE_HANDLE -DPTBMODULE_PsychVulkanCore -ICommon/Base -ILinux/Base -ICommon/PsychVulkanCore "Linux/Base/*.c" "Common/Base/*.c" "Common/PsychVulkanCore/*.c" -lc -lLexActivator -lrt -ldl -lX11 -lvulkan
+        mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread', '-outdir', '../Projects/Linux/build/', '-output', 'PsychVulkanCore', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-DPTBMODULE_PsychVulkanCore', '-ICommon/Base', '-ILinux/Base', '-ICommon/PsychVulkanCore', 'Linux/Base/*.c', 'Common/Base/*.c', 'Common/PsychVulkanCore/*.c', '-lc', '-lrt', '-ldl', '-lX11', '-lvulkan');
     catch
         disp(psychlasterror);
     end
@@ -159,7 +177,7 @@
 if mode==16
     % Build PsychOpenXRCore.mexa64:
     try
-        mex 'LDFLAGS=$LDFLAGS -Wl,-rpath,\$ORIGIN/PsychPlugins,-rpath,\$ORIGIN/PsychPlugins/Intel64' -L../../Psychtoolbox/PsychBasic/PsychPlugins -L../../Psychtoolbox/PsychBasic/PsychPlugins/Intel64 CFLAGS='$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread' -outdir ../Projects/Linux/build/ -output PsychOpenXRCore -largeArrayDims -DMEX_DOUBLE_HANDLE -D_GNU_SOURCE -DPTBMODULE_PsychOpenXRCore -ICommon/Base -ILinux/Base -ICommon/PsychOpenXRCore -ICommon/PsychOpenXRCore/nanopb "Linux/Base/*.c" "Common/Base/*.c" "Common/PsychOpenXRCore/*.c" "Common/PsychOpenXRCore/nanopb/*.c" -lc -lLexActivator -lrt -ldl -lopenxr_loader
+        mex('CFLAGS=$CFLAGS -fPIC -std=gnu99 -fexceptions -pthread', '-outdir', '../Projects/Linux/build/', '-output', 'PsychOpenXRCore', '-largeArrayDims', '-DMEX_DOUBLE_HANDLE', '-DPSYCH_NOLM', '-D_GNU_SOURCE', '-DPTBMODULE_PsychOpenXRCore', '-ICommon/Base', '-ILinux/Base', '-ICommon/PsychOpenXRCore', '-ICommon/PsychOpenXRCore/nanopb', 'Linux/Base/*.c', 'Common/Base/*.c', 'Common/PsychOpenXRCore/*.c', 'Common/PsychOpenXRCore/nanopb/*.c', '-lc', '-lrt', '-ldl', '-lopenxr_loader');
     catch
         disp(psychlasterror);
     end
@@ -167,3 +185,13 @@
 end
 
 return;
+
+% Ask pkg-config for include flags instead of hardcoding FHS and multiarch
+% paths, so the build works outside of Debian/Ubuntu.
+function flags = pkgconfig_iflags(pkgs)
+[rc, out] = system(['pkg-config --cflags-only-I ' pkgs]);
+if rc ~= 0
+    error('pkg-config failed for: %s', pkgs);
+end
+flags = strsplit(strtrim(out));
+flags = flags(~cellfun('isempty', flags));
